File: ru.kmorozov.gbd.desktop/src/main/java/ru/kmorozov/gbd/desktop/output/progress/ProcessStatus.java
package ru.kmorozov.gbd.desktop.output.progress;

import ru.kmorozov.gbd.core.config.SystemConfigs;
import ru.kmorozov.gbd.logger.progress.IProgress;
import ru.kmorozov.gbd.desktop.gui.MainBookForm;

import javax.swing.*;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Created by km on 22.12.2015.
 */
public class ProcessStatus implements IProgress {

    private final AtomicInteger value = new AtomicInteger(0);
    private JProgressBar prBar;
    private int maxValue;

    public ProcessStatus(int maxValue) {
        this.maxValue = maxValue;

        this.start();
    }

    public ProcessStatus() {

    }

    @Override
    public void resetMaxValue(int maxValue) {
        this.maxValue = maxValue;

        this.start();
    }

    @Override
    public int inc() {
        return this.value.incrementAndGet();
    }

    public int get() {
        return this.value.get();
    }

    @Override
    public int incrementAndProgress() {
        return 0 == this.maxValue ? 0 : Math.round((float) Math.min(this.inc() * 100 / this.maxValue, 100));
    }

    private void start() {
        if (SystemConfigs.isConsoleMode()) return;

        this.prBar = new JProgressBar();
        this.prBar.setMinimum(0);
        this.prBar.setMinimum(this.maxValue);
        this.prBar.setIndeterminate(false);

        SwingUtilities.invokeLater(() -> {
            MainBookForm.getINSTANCE().getProgressPanel().add(this.prBar);
            SwingUtilities.updateComponentTreeUI(MainBookForm.getINSTANCE().getProgressPanel());
        });
    }

    @Override
    public void finish() {
        if (SystemConfigs.isConsoleMode()) return;

        SwingUtilities.invokeLater(() -> {
            MainBookForm.getINSTANCE().getProgressPanel().remove(this.prBar);
            SwingUtilities.updateComponentTreeUI(MainBookForm.getINSTANCE().getProgressPanel());
        });
    }

    public JProgressBar getProgressBar() {
        return this.prBar;
    }

    @Override
    public IProgress getSubProgress(int maxValue) {
        return new ProcessStatus(maxValue);
    }
}

File: ru.kmorozov.gbd.desktop/src/main/java/ru/kmorozov/gbd/desktop/output/progress/ProcessStatus.kt
package ru.kmorozov.gbd.desktop.output.progress

import ru.kmorozov.gbd.core.config.SystemConfigs
import ru.kmorozov.gbd.logger.progress.IProgress
import ru.kmorozov.gbd.desktop.gui.MainBookForm

import javax.swing.*
import java.util.concurrent.atomic.AtomicInteger

/**
 * Created by km on 22.12.2015.
 */
class ProcessStatus : IProgress {

    private val value = AtomicInteger(0)
    var progressBar: JProgressBar? = null
        private set
    private var maxValue: Int = 0

    constructor(maxValue: Int) {
        this.maxValue = maxValue

        start()
    }

    constructor() {

    }

    override fun resetMaxValue(maxValue: Int) {
        this.maxValue = maxValue

        start()
    }

    override fun inc(): Int {
        return value.incrementAndGet()
    }

    fun get(): Int {
        return value.get()
    }

    override fun incrementAndProgress(): Int {
        return if (0 == maxValue) 0 else Math.round(Math.min(inc() * 100 / maxValue, 100).toFloat())
    }

    private fun start() {
        if (SystemConfigs.isConsoleMode) return

        progressBar = JProgressBar()
        progressBar!!.minimum = 0
        progressBar!!.minimum = maxValue
        progressBar!!.isIndeterminate = false

        SwingUtilities.invokeLater {
            MainBookForm.instance!!.progressPanel!!.add(progressBar)
            SwingUtilities.updateComponentTreeUI(MainBookForm.instance!!.progressPanel!!)
        }
    }

    override fun finish() {
        if (SystemConfigs.isConsoleMode) return

        SwingUtilities.invokeLater {
            MainBookForm.instance!!.progressPanel!!.remove(progressBar!!)
            SwingUtilities.updateComponentTreeUI(MainBookForm.instance!!.progressPanel!!)
        }
    }

    override fun getSubProgress(maxValue: Int): IProgress {
        return ProcessStatus(maxValue)
    }
}
